functions/go/set-labels/transformer.go

```diff
@@ -0,0 +1,301 @@
+package main
+
+import (
+	"sort"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+)
+
+type FieldPath []string
+var (
+	selectorPath = FieldPath{"selector", "matchLabels"}
+	metaLabelsPath = FieldPath{"metadata", "labels"}
+	podSelectorPath = FieldPath{"podSelector", "matchLabels"}
+
+)
+
+
+var _ fn.Runner = &SetLabels{}
+
+// SetLabels supports the set-labels workflow, it uses Config to parse functionConfig, Transform to change the labels
+type SetLabels struct {
+	// labels is the desired labels
+	Labels map[string]string `json:"labels,omitempty"`
+	count int
+}
+
+// Transform updates the labels in the right path using GVK filter and other configurable fields
+func (p *SetLabels) Run(_ *fn.Context, fnConfig *fn.KubeObject, items fn.KubeObjects, results *fn.Results) bool {
+	objects := items.WhereNot(fn.IsLocalConfig)
+	if objects.Len() == 0 || objects[0] == nil {
+		results.Infof("no input resources")
+		return true
+	}
+	// A workaround since kpt creates a FunctionConfig placeholder if users don't provide the functionConfig.
+	if fn.EmptyFunctionConfig(fnConfig) {
+		return false
+	}
+	if len(p.Labels) == 0 {
+		results.Warningf("no `labels` arguments are given in FunctionConfig")
+		return true
+	}
+	p.count = 0
+	for _, o := range objects {
+		err := func() error {
+			if err := p.setLabelsInMeta(&o.SubObject); err != nil {
+				return err
+			}
+			if err := p.setLabelsInSelector(o); err != nil {
+				return err
+			}
+			if containPodSpec(o) {
+				pod, found, err := o.NestedSubObject("spec", "template")
+				if err != nil || !found {
+					return err
+				}
+				if err := p.setLabelsInPod(&pod); err != nil {
+					return err
+				}
+			}
+			if containJobSpec(o) {
+				job, found, err := o.NestedSubObject("spec", "jobTemplate")
+				if err != nil || !found {
+					return err
+				}
+				if err := p.setLabelsInJob(&job); err != nil {
+					return err
+				}
+			}
+			if containVolume(o) {
+				volume, found, err := o.NestedSubObject("spec", "volumeClaimTemplates")
+				if err != nil || !found {
+					return err
+				}
+				if err := p.setLabelsInVolume(&volume); err != nil {
+					return err
+				}
+			}
+
+			if containNetworkPolicy(o) {
+				if err := p.setNetworkPolicyRule(o); err != nil {
+					return err
+				}
+			}
+			return nil
+		}()
+		if err != nil {
+			results.ErrorE(err)
+		}
+	}
+	results.Infof("set %v labels in total", p.count)
+	return results.ExitCode() != 1
```

Comment:
we can probably have a helper `results.HasErrors()` returning true/false because `ExitCode()` left me wondering how is that determined.

Reply:
Sounds good! I actually have a draft changes on similar functions.